Rails 4 changes

dLobatog opened the pull request.

ohadlevy commented on Gemfile:
```diff
@@ -27,6 +27,8 @@ gem 'gettext_i18n_rails', '~> 1.0'
 gem 'turbolinks', '~> 2.5'
 gem 'logging', '>= 1.8.0', '< 3.0.0'
+gem 'activerecord-session_store'
+gem 'rails-observers'
```
i wonder if we should not move the only observer that we have back into the host class.

ohadlevy commented on Gemfile:
```diff
@@ -4,22 +4,22 @@ require File.expand_path('../lib/regexp_extensions', FOREMAN_GEMFILE)
 source 'https://rubygems.org'
 
-gem 'rails', '3.2.21'
+gem 'rails', '4.1.5'
```
imho this should be latest e.g. 4.1.12?

ohadlevy commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -49,7 +49,7 @@ function fix_template_context(content, context) {
 function fix_template_names(content, assoc, new_id) {
   var regexp  = new RegExp('new_' + assoc, 'g');
-  return content.replace(regexp, "new_" + new_id);
+  return content.replace(regexp, new_id);
```
just for background, why is this part of this PR ?

ohadlevy commented on app/controllers/unattended_controller.rb:
```diff
@@ -45,8 +45,8 @@ def built
 
-    template = ProvisioningTemplate.find(params['id'])
-    @host = Hostgroup.find(params['hostgroup'])
+    template = ProvisioningTemplate.find(params['id']) || ProvisioningTemplate.where(:name => params[:id]).first
```
friendly id does not work here?

ohadlevy commented on app/controllers/unattended_controller.rb:
```diff
@@ -56,9 +56,9 @@ def template
-  TemplateKind.all.each do |kind|
-    define_method kind.name do
-      render_template kind.name
+  TemplateKind.all.map(&:name).each do |name|
```
pluck name?

ohadlevy commented on app/helpers/application_helper.rb:
```diff
@@ -32,6 +32,13 @@ def link_to(*args, &block)
   end
 
+  def link_to_function(name, function, html_options = {})
+    onclick = "#{"#{html_options[:onclick]}; " if html_options[:onclick]}#{function}; return false;"
```
does this line works? can you please rewrite it please to be readable as well? :)

ohadlevy commented on app/helpers/application_helper.rb:
```diff
@@ -32,6 +32,13 @@ def link_to(*args, &block)
     end
   end
 
+  def link_to_function(name, function, html_options = {})
```
why are we rewrite link_to_function? is it RIP ?

ohadlevy commented on app/mailers/host_mailer.rb:
```diff
@@ -30,7 +30,9 @@ def summary(options = {})
       }
 
-      mail(:to => user.mail, :subject => subject)
+      mail(:to   => user.mail,
```
nitpick: extra whitespaces?

ohadlevy commented on app/models/auth_sources/auth_source_ldap.rb:
```diff
@@ -21,13 +21,11 @@ class AuthSourceLdap < AuthSource
   encrypts :account_password
 
-  validates :host, :presence => true, :length => {:maximum => 60}, :allow_nil => true
+  validates :host, :presence => true, :length => {:maximum => 60}, :if => Proc.new{|auth| !auth.new_record?}
```
why only if? shouldnt we validate on create&&update?

ohadlevy commented on app/models/compute_attribute.rb:
```diff
@@ -12,6 +11,9 @@ class ComputeAttribute < ActiveRecord::Base
 
   def method_missing(method, *args, &block)
+    #apparently, rails 4 defines attribute methods on method missing.
+    return super if method.to_s[-1]=="="
```
nitpick: whitespace

ohadlevy commented on app/models/concerns/counter_cache_fix.rb:
```diff
@@ -7,7 +7,7 @@ module CounterCacheFix
 
   included do
-    after_update :update_counter_caches
+    #after_update :update_counter_caches
```
commented out code?

ohadlevy commented on app/models/concerns/host_common.rb:
```diff
@@ -114,18 +124,19 @@ def crypt_root_pass
-                              else
-                                unencrypted_pass.starts_with?("$")
-                              end
+      if PasswordCrypt.passw_crypt("test_this").match('^\$\d+\$.+\$.+')
```
nitpick: why change here?

ohadlevy commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -81,7 +81,8 @@ def dhcp_attrs
 
   def queue_dhcp
-    return unless (dhcp? or (old and old.dhcp?)) and orchestration_errors?
+    return unless orchestration_errors?
```
why swap the order?

ohadlevy replied:
imho outside the scope of this pr

ohadlevy commented on app/models/concerns/taxonomix.rb:
```diff
@@ -3,12 +3,14 @@ module Taxonomix
-    has_many :organizations, :through => taxonomy_join_table, :source => :taxonomy,
-             :conditions => "taxonomies.type='Organization'", :validate => false
+    taxonomy_join_table = :taxable_taxonomies
+    has_many taxonomy_join_table.to_sym, :dependent => :destroy, :as => :taxable
```
its already a sym, no need to change it again

ohadlevy commented on app/models/host/managed.rb:
```diff
@@ -21,7 +21,7 @@ class Host::Managed < Host::Base
     matcher = /(\s*(?:(?:user\.[a-z]+)|owner)\s*[=~])\s*(\S*)\s*\z/
     matches = matcher.match(query)
-    output = super(query, opts)
+    output = super
```
whats super without query and opts? I dont think this could work?

ohadlevy commented on app/models/host/managed.rb:
```diff
@@ -633,7 +632,7 @@ def root_pass
   def clone
     # do not copy system specific attributes
     host = self.deep_clone(:include => [:host_config_groups, :host_classes, :host_parameters],
-                           :except  => [:name, :mac, :ip, :uuid, :certname, :last_report])
```
is this in develop? maybe a separate PR ?

ohadlevy commented on app/models/hostgroup.rb:
```diff
@@ -14,7 +13,7 @@ class Hostgroup < ActiveRecord::Base
   validates_lengths_from_database :except => [:name]
   before_destroy EnsureNotUsedBy.new(:hosts)
-  has_many :hostgroup_classes
+  has_many :hostgroup_classes, :dependent => :destroy
```
also maybe a separate PR to develop?

ohadlevy commented on app/models/hostgroup.rb:
```diff
@@ -162,8 +167,8 @@ def inherited_params(include_source = false)
-    hp.update organization.parameters(include_source) if SETTINGS[:organizations_enabled] && organization
-    hp.update location.parameters(include_source)     if SETTINGS[:locations_enabled] && location
+    #hp.update organization.parameters(include_source) if SETTINGS[:organizations_enabled] && organization
+    #hp.update location.parameters(include_source)     if SETTINGS[:locations_enabled] && location
```
commented out

ohadlevy commented on app/models/hostgroup_class.rb:
```diff
@@ -7,9 +7,9 @@ class HostgroupClass < ActiveRecord::Base
-  validates :hostgroup_id, :presence => true
+  #disabled because when created with hostgroup, fails validation
+  #TODO: find a way to fix this
+  #validates :hostgroup_id, :presence => true
```
?

ohadlevy commented on app/models/medium.rb:
```diff
@@ -19,7 +17,7 @@ class Medium < ActiveRecord::Base
   validates :name, :uniqueness => true, :presence => true
   validates :path, :uniqueness => true, :presence => true,
-                   :format => { :with => /^(http|https|ftp|nfs):\/\//,
+                   :format => { :with => /\A(http|https|ftp|nfs):\/\//,
```
separate PR ?

ohadlevy commented on app/models/ptable.rb:
```diff
@@ -51,7 +47,7 @@ def self.base_class
 
   def self.template_includes
-    super + [:operatingsystems]
+    super #+ [:operatingsystems]
```
?

ohadlevy commented on app/models/smart_proxy.rb:
```diff
@@ -19,7 +16,7 @@ class SmartProxy < ActiveRecord::Base
   has_many :puppet_ca_hostgroups, :class_name => 'Hostgroup', :foreign_key => 'puppet_ca_proxy_id'
   has_many :realms,                                           :foreign_key => 'realm_proxy_id'
-  URL_HOSTNAME_MATCH = %r{^(?:http|https):\/\/([^:\/]+)}
+  URL_HOSTNAME_MATCH = %r{\A(?:http|https):\/\/([^:\/]+)}
```
same separate PR ?

ohadlevy commented on app/models/user.rb:
```diff
@@ -48,15 +47,17 @@ class User < ActiveRecord::Base
-    if (hidden = AuthSourceHidden.all).present?
-      where("#{self.table_name}.auth_source_id <> ?", hidden)
+    if (hidden = AuthSourceHidden.all.select(:id)).present?
+      where("#{self.table_name}.auth_source_id <> (?)", hidden).references(:users)
```
cant this be rewritten to one query instead of two?

ohadlevy commented on app/models/user.rb:
```diff
@@ -384,7 +385,7 @@ def self.random_password(size = 16)
 
   def expire_topbar_cache(sweeper)
-    sweeper.expire_fragment(TopbarSweeper.fragment_name(id))
+    sweeper.expire_fragment(TopbarSweeper.fragment_name(id)) rescue nil
```
???

ohadlevy commented on app/models/user_mail_notification.rb:
```diff
@@ -1,10 +1,8 @@
   belongs_to :mail_notification
 
-  validates :user_id, :presence => true
+  #validates :user_id, :presence => true
```
?

ohadlevy commented on app/models/user_role.rb:
```diff
@@ -22,8 +22,9 @@ class UserRole < ActiveRecord::Base
-  validates :owner_id, :presence => true, :uniqueness => {:scope => [:role_id, :owner_type],
+  validates :owner_id, :uniqueness => {:scope => [:role_id, :owner_type],
                                                           :message => N_("has this role already")}
+
```
nitpick: whiatespace

ohadlevy commented on app/services/authorizer.rb:
```diff
@@ -80,6 +76,7 @@ def build_filtered_scope_components(resource_class, all_filters, options)
 
     result[:where] << { id: base_ids } if @base_collection.present?
     return result if all_filters.any?(&:unlimited?)
+    # return resource_class.all if all_filters.any?(&:unlimited?)
```
??

ohadlevy commented on app/services/fact_importer.rb:
```diff
@@ -49,7 +49,9 @@ def fact_name_class
-    to_delete = host.fact_values.joins(:fact_name).where('fact_names.name NOT IN (?)', facts.keys)
+    #in rails4 join makes the returned set read only, so let's collect the ids and fetch everything again. :(
+    to_delete_ids = host.fact_values.joins(:fact_name).where('fact_names.name NOT IN (?)', facts.keys).pluck(:id)
+    to_delete = host.fact_values.where(id: to_delete_ids)
```
AFAIK there is a way to mark the results as RW

ohadlevy commented on app/services/tax_host.rb:
```diff
@@ -155,7 +155,7 @@ def non_inherited_ids(v1 = self.selected_ids, v2 = self.inherited_ids)
   def user_ids(hosts = self.hosts)
     #TODO: when migrating to rails 3.1+ switch to inner select on users.
-    User.unscoped.joins(:direct_hosts).where({ :hosts => { :id => hosts }, :users => { :admin => false } }).pluck('DISTINCT users.id')
+    User.unscoped.joins(:direct_hosts).where({ :hosts => { :id => hosts.map(&:id) }, :users => { :admin => false } }).pluck('DISTINCT users.id')
```
or just {:hosts => hosts} ?

ohadlevy commented on app/views/settings/index.html.erb:
```diff
@@ -2,6 +2,7 @@
 <% title_actions documentation_button('3.5.2ConfigurationOptions') %>
 <ul class="nav nav-tabs" data-tabs="tabs">
   <% @settings.group_by(&:category).each do |category, setting| %>
+    <% next if category.nil? %>
```
separate PR ?

ohadlevy commented on app/views/settings/index.html.erb:
```diff
@@ -9,6 +10,7 @@
 </ul>
 <div class="tab-content">
   <% @settings.group_by(&:category).each do |category, setting| %>
+    <% next if category.nil? %>
```
same as above

ohadlevy commented on app/views/templates/_form.html.erb:
```diff
@@ -1,6 +1,6 @@
 <%= include_javascript %>
-<% url = @template.persisted? ? public_send("#{@type_name_singular}_path", @template) : public_send("#{@type_name_singular}_path", @resource_class.new) %>
-<%= form_for @template, :url => url, :html => { :multipart => true, :onsubmit => 'submit_code();' } do |f| %>
+<%# url = @template.persisted? ? public_send("#{@type_name_singular}_path", @template) : public_send("#{@type_name_singular}_path", @resource_class.new) %>
```
?

ohadlevy commented on config/application.rb:
```diff
@@ -196,6 +196,14 @@ class Application < Rails::Application
   end
 
+begin
+  require 'rack/openid'
```
initializer? I can't remember, but what actually uses this?

ohadlevy commented on config/initializers/backtrace_silencers.rb:
```diff
@@ -4,10 +4,10 @@
-  line !~ Rails::BacktraceCleaner::APP_DIRS_PATTERN &&
-      !(Foreman::Plugin.all.any? { |plugin| line.include?(plugin.name) })
-end
+#Rails.backtrace_cleaner.add_silencer do |line|
```
?

ohadlevy commented on config/routes.rb:
```diff
@@ -8,7 +8,7 @@
 
   #ENC requests goes here
-  match "node/:name" => 'hosts#externalNodes', :constraints => { :name => /[^\.][\w\.-]+/ }
+  get "node/:name" => 'hosts#externalNodes', :constraints => { :name => /[^\.][\w\.-]+/ }
```
separate PR ?

ohadlevy commented on db/migrate/20110616080444_add_look_up_key_id_to_puppet_class.rb:
```diff
@@ -11,13 +11,13 @@ def up
     add_column :lookup_keys, :validator_rule, :string
     rename_column :lookup_values, :priority, :match
-    add_index :lookup_values, :match
+    #add_index :lookup_values, :match
```
?

ohadlevy commented on db/migrate/20140901115249_add_request_uuid_to_audits.rb~HEAD:
```diff
@@ -0,0 +1,10 @@
+class AddRequestUuidToAudits < ActiveRecord::Migration
```
oops?

ohadlevy commented on lib/foreman.rb:
```diff
@@ -13,6 +13,7 @@ def self.is_uuid?(str)
   end
 
   def self.in_rake?(rake_task = nil)
+    return false if Rails.env.test?
```
is this on purpose?

dLobatog committed:
Fixes #7230, #12021 - Upgrade to Rails 4.1.5

This commits upgrades Rails to Rails 4.1.5. See a description of the
changes included here, and go to the pull request in GitHub to see more
detailed explanations:

* Update gems to a Rails 4 compatible version, including dependencies
* Fix counter cache columns
* Remove conditions, order, limit from has_many relations
* Remove test runner
    On minitest 5, the runner API was deprecated, so our custom test
    runner is no longer working.

* Remove useless add_index on lookup_values match
    An index is added previously on lookup_values :priority, and on Rails 4
    rename_column changes 'priority' to 'match' already changes the name of
    the index.

* Alias assert_include to assert_includes.
* Expire topbar cache
* Subclass ApplicationMailer in test to avoid 'missing template' error
* Fixes #12021 - Use .to_param to find parent object in LookupKeys controller
    On Rails 4, .find will not default to .friendly.find so find_hostgroup,
    find_environment and find_host will fail for non numeric IDs. However, we can
    use from_param to find these objects. This is something we can do both on
    Rails 3 and Rails 4, whereas using .friendly isn't an option until friendly_id
     5.0, which depends on Rails 4.

* Use explicit friendly ID search.
* find_common finder uses from_param -> friendly -> find
    The finder needed to be refactored because with the new Friendly ID we
    have to use .friendly explicitely. It currently follows the strategy of
    searching like this:
      - from_param -> .friendly -> regular find

* Remove RecordNotFound errors from parameterizable in rails 4
* Protect attributes using attr_accessible and protected_attributes
* url_for doesn't append port when protocol is specified
    Some of our tests are checking if port 80 or 443 is included in the URL
    generated by lib/foreman/renderer.rb. However url_for has changed in
    Rails 4 and now it does not append the port even if explicitly added.

    If the protocol is specified, as it is in this case, then the url
    generated will just have the protocol (http or https) but not the port
    (80 or 443)

* Validate object instead of _id column in join tables
    On Rails 4 such validations will fail when we try to create objects
    without explicitely assigning the id.
    Puppetclass.new(:config_group => config_group) would fail even for a
    valid config group, :config_group_id => config_group.id should be used
    instead.

    To avoid that, we validate the object, not the ID column

* Rails 4 non backwards compatible syntax changes
    Changes that have to deal with how some of the internal Rails objects
    need a new syntax, like generating routes, exceptions, form builders.
    None of these are Rails 3 compatible.

* Fixes #12199 - rails 4 migration errors
    There are actually two parts to the issue of not being able to migrate
    with this branch. The first is
    https://gist.github.com/eLobato/0f5db50b5c93cc6c277c and can be
    temporarily fixed with rails/sass-rails#136 (comment) I am not sure of a
    long term fix for that atm.

    Once that is patched, this error
    https://gist.github.com/johnpmitsch/96e5ba3629890931193a happens on a
    migration. This will fix that error by changing the Migrator class
    initializer arguments which have changed from rails 3 to rails 4

* Refactor fact value test to not modify user_roles directly
* Return external usergroupsas an array
* Use eager_load to preload associations to be used in where

Fixes #7230, #12021 - Upgrade to Rails 4.1.5

This commits upgrades Rails to Rails 4.1.5. See a description of the
changes included here, and go to the pull request in GitHub to see more
detailed explanations:

* Update gems to a Rails 4 compatible version, including dependencies
* Fix counter cache columns
* Remove conditions, order, limit from has_many relations
* Remove test runner
    On minitest 5, the runner API was deprecated, so our custom test
    runner is no longer working.

* Remove useless add_index on lookup_values match
    An index is added previously on lookup_values :priority, and on Rails 4
    rename_column changes 'priority' to 'match' already changes the name of
    the index.

* Alias assert_include to assert_includes.
* Expire topbar cache
* Subclass ApplicationMailer in test to avoid 'missing template' error
* Fixes #12021 - Use .to_param to find parent object in LookupKeys controller
    On Rails 4, .find will not default to .friendly.find so find_hostgroup,
    find_environment and find_host will fail for non numeric IDs. However, we can
    use from_param to find these objects. This is something we can do both on
    Rails 3 and Rails 4, whereas using .friendly isn't an option until friendly_id
     5.0, which depends on Rails 4.

* Use explicit friendly ID search.
* find_common finder uses from_param -> friendly -> find
    The finder needed to be refactored because with the new Friendly ID we
    have to use .friendly explicitely. It currently follows the strategy of
    searching like this:
      - from_param -> .friendly -> regular find

* Remove RecordNotFound errors from parameterizable in rails 4
* Protect attributes using attr_accessible and protected_attributes
* url_for doesn't append port when protocol is specified
    Some of our tests are checking if port 80 or 443 is included in the URL
    generated by lib/foreman/renderer.rb. However url_for has changed in
    Rails 4 and now it does not append the port even if explicitly added.

    If the protocol is specified, as it is in this case, then the url
    generated will just have the protocol (http or https) but not the port
    (80 or 443)

* Validate object instead of _id column in join tables
    On Rails 4 such validations will fail when we try to create objects
    without explicitely assigning the id.
    Puppetclass.new(:config_group => config_group) would fail even for a
    valid config group, :config_group_id => config_group.id should be used
    instead.

    To avoid that, we validate the object, not the ID column

* Rails 4 non backwards compatible syntax changes
    Changes that have to deal with how some of the internal Rails objects
    need a new syntax, like generating routes, exceptions, form builders.
    None of these are Rails 3 compatible.

* Fixes #12199 - rails 4 migration errors
    There are actually two parts to the issue of not being able to migrate
    with this branch. The first is
    https://gist.github.com/eLobato/0f5db50b5c93cc6c277c and can be
    temporarily fixed with rails/sass-rails#136 (comment) I am not sure of a
    long term fix for that atm.

    Once that is patched, this error
    https://gist.github.com/johnpmitsch/96e5ba3629890931193a happens on a
    migration. This will fix that error by changing the Migrator class
    initializer arguments which have changed from rails 3 to rails 4

* Refactor fact value test to not modify user_roles directly
* Return external usergroupsas an array
* Use eager_load to preload associations to be used in where
* Auto detect jenkins rake task in application.rb and set test RAILS_ENV
  in that case
* Reference previously undigested assets in CSS and js as digested using
  asset-url or image-url
* Adapted sprockets manifest.json instead of manifest.yml (ehelms)
